pkg/controller/api.go

```diff
@@ -67,39 +74,109 @@ func NewAPICallbacks(m ctrl.Manager, of xpresource.ManagedKind) *APICallbacks {
 	return &APICallbacks{
 		kube:           m.GetClient(),
 		newTerraformed: nt,
+		EventHandler: &eventHandler{
+			mu:          &sync.Mutex{},
+			rateLimiter: workqueue.DefaultControllerRateLimiter(),
+		},
 	}
 }
 
 // APICallbacks providers callbacks that work on API resources.
 type APICallbacks struct {
+	EventHandler *eventHandler
+
 	kube           client.Client
 	newTerraformed func() resource.Terraformed
 }
 
-// Apply makes sure the error is saved in async operation condition.
-func (ac *APICallbacks) Apply(name string) terraform.CallbackFn {
+func (ac *APICallbacks) callbackFn(name, op string) terraform.CallbackFn {
 	return func(err error, ctx context.Context) error {
 		nn := types.NamespacedName{Name: name}
 		tr := ac.newTerraformed()
 		if kErr := ac.kube.Get(ctx, nn, tr); kErr != nil {
-			return errors.Wrap(kErr, errGet)
+			return errors.Wrapf(kErr, errGetFmt, tr.GetObjectKind().GroupVersionKind().String(), name, op)
 		}
 		tr.SetConditions(resource.LastAsyncOperationCondition(err))
 		tr.SetConditions(resource.AsyncOperationFinishedCondition())
-		return errors.Wrap(ac.kube.Status().Update(ctx, tr), errStatusUpdate)
+		return errors.Wrapf(ac.kube.Status().Update(ctx, tr), errUpdateStatusFmt, tr.GetObjectKind().GroupVersionKind().String(), name, op)
 	}
 }
 
-// Destroy makes sure the error is saved in async operation condition.
-func (ac *APICallbacks) Destroy(name string) terraform.CallbackFn {
+// Create makes sure the error is saved in async operation condition.
+func (ac *APICallbacks) Create(name string) terraform.CallbackFn {
```

Comment:
Done. Refactored the requeue logic into the `controller.callbackFn` function. In the context of `controller.CallbackProvider`, I think it's good to have an interface reflecting the logical functions (async create/update/delete) instead of the implementation specific functions (both create and update are implemented via a `terraform apply`). In the feature, for example, we may want to explore different service classes for the create & create operations, e.g., we may want to prioritize async create operations over update operations if we want to optimize certain metrics for example.